### ISSUE TITLE
JWT from hub's OPTIONS request

### ISSUE THREAD

**Issue by SGStino** [aspnet/SignalR#994]
I was looking into getting the JWT token in the header when creating the initial connection, and this seems to work as far as the .NET client goes.
However, I can't find anything server-side that contains that token.

I know that a recommended way to get the JWT to the server is by using the query string (#130). But I'm not sure if that's the best place to leave a token. (max query length and big tokens, logs, etc.)

```http
OPTIONS http://192.168.1.10:5000/hubs/events HTTP/1.1
User-Agent: Microsoft.AspNetCore.Sockets.Client.Http/1.0.0-alpha
Authorization: Bearer eyJ0e...
Host: 192.168.1.171:56419
Content-Length: 0
```
```http
HTTP/1.1 200 OK
Date: Fri, 06 Oct 2017 17:16:08 GMT
Content-Type: application/json
Server: Kestrel
Content-Length: 125
Allow: GET,POST,OPTIONS
Request-Context: appId=cid-v1:b91b490a-68aa-4451-829d-55b1076112c2
{"connectionId":"63284d81-88fa-4200-95e9-8171213cacc3","availableTransports":["WebSockets","ServerSentEvents","LongPolling"]}
```
```http
GET http://192.168.1.10:5000/hubs/events?id=63284d81-88fa-4200-95e9-8171213cacc3 HTTP/1.1
Connection: Upgrade
Upgrade: websocket
Sec-WebSocket-Key: Xsl/8XU4BT4NXsV3+MVDPA==
Sec-WebSocket-Version: 13
Host: 192.168.1.171:56419
```
```http
HTTP/1.1 101 Switching Protocols
Connection: Upgrade
Date: Fri, 06 Oct 2017 17:16:08 GMT
Server: Kestrel
Upgrade: websocket
Request-Context: appId=cid-v1:b91b490a-68aa-4451-829d-55b1076112c2
Sec-WebSocket-Accept: dlqwfi6JHEQZJsRq8xbDTcFv9kg=
```

The initial `OPTIONS` call is authenticated, and it creates the connection id, which can only be used once if I'm not mistaken. This means, that at this point, the server knows the mapping from the connection id to the ClaimsPrincipal of the user. Without the JWT token ever appearing in any logs. So the token is basically swapped out for a one-time use password (the connection id).

So in theory, if this could work, this should be more secure than the query parameter unless I'm missing something vital?

The first OPTIONS call and the GET with the connection upgrade need to be handled by the same server instance? Because another node won't know the connection id?

And for the long-polling transport it could just fall back to providing the Authorize header in each request.

**Comment by SGStino** [comment]
My initial guess would be to change [HttpConnectionDispatcher.cs#L335](https://github.com/aspnet/SignalR/blob/3c5d283689879ed50f5a249146c6c3b559cef53a/src/Microsoft.AspNetCore.Sockets.Http/HttpConnectionDispatcher.cs#L335)
To record the user's ClaimsPrincipal from that HttpContext inside The ConnectionContext?

**Comment by moozzyk** [comment]
Connection id is not required for webSockets and we are actually thinking about removing the OPTIONS requests if the user select webSockets transport.

**Comment by SGStino** [comment]
That makes sense ... only required when negotiating, and the client application usually knows what the server's capable off (all 3 transports). 

Could make load balancing scenarios for WebSockets-only setups easier too I guess? 
But this means I'm back to query strings.

**Comment by moozzyk** [comment]
Correct - the negotiation is really required when using websockets since once you connect the connection is persistent. It will simplify load balancing scenarios but will (along with removing the negotiate message in some specific cases/setup) also make it easier to connect to the server with bare WebSocket (i.e. without using a dedicated SignalR client).

**Comment by moozzyk** [comment]
I merged a change (0bafb304c2707739578e635bc9e00d0cd36420a7 18a65496b7adab684a9af2ac255c7128699c96ee) today that adds an API that allows to pass a JWT token when starting a connection. When using the C# client the JWT token will be passed in the header. When using the TS client the token will be passed in the header when sending an HTTP request with XmlHttpRequest (i.e. for negotiate, long polling - send/poll request, server sent events - send request) or in the query string if the underlying API does not allow setting headers (webSockets, eventSource used for the server-to-client channel in the ServerSentEvents transport) as the `signalRTokenHeader` queryString parameter. When using the TS client it is required to copy the token to the MessageReceivedContext.Token as shown in the [sample](https://github.com/aspnet/SignalR/blob/dev/samples/JwtSample/Startup.cs#L49-L62).
